README
 MS-DOS Shell Version 2.2	README				December 1993

 MS-DOS SHELL - Copyright (c) 1990,3 Data Logic Limited and Charles Forsyth

 This code is based on (in part) the shell program written by Charles
 Forsyth and is subject to the following copyright restrictions:

 1.  Redistribution and use in source and binary forms are permitted
     provided that the above copyright notice is duplicated in the
     source form and the copyright notice in file sh6.c is displayed
     on entry to the program.

 2.  The sources (or parts thereof) or objects generated from the
     sources (or parts of sources) cannot be sold under any circumstances.

    $Header: /usr/users/istewart/src/shell/sh2.2/Release/RCS/ReadMe,v 2.2 1993/12/03 13:32:52 istewart Exp $

    $Log: ReadMe,v $
	Revision 2.2  1993/12/03  13:32:52  istewart
	Release 2.2

	Revision 2.1  1992/12/14  11:14:32  istewart
	BETA 215 Fixes and 2.1 Release

	Revision 2.0  1992/04/13  17:40:33  Ian_Stewartson
	MS-Shell 2.0 Baseline release

	
________________________________________________________________________________

This is an implementation of the Unix Shell for MSDOS.  As far as possible it
is compatible with the System V.4 program sh(1) with ksh(1) extension.  The
following differences are noted:

1)  Background or asynchronous commands are not supported under MSDOS.  Some
    support is provided under OS/2, but it is not as full as UNIX.  You can
    only run one command and not a pipeline in background.

2)  Certain internal commands which have no equivalent MSDOS or OS/2 supported
    functionality support (ulimit, time etc) are not provided.

3)  Command hashing and accounting are not supported.

4)  Some of the ksh functionality has not yet been implemented or made ksh
    compatible (see Notes file).

5)  The Shell uses all variables starting with a ~ (tilde) internally and will
    not allow you to display them.  I don't think this is a difference from
    the user's view, just internally.

The Shell has been tested mainly under MSDOS 3.3, 4.01, 5.2, 6.0, 6.2 and OS/2
1.3, 2.0 and 2.1.  It has exposed some shortcomings in the C6.0 compiler and
MSDOS library.  As such, using C5.1 and its library are prefered when
re-building the shell for MSDOS.

The following enhancements have been made for the MSDOS environment.  These
enhancements are described in the appropriate section of the manual pages.

1)  Under MSDOS (16-bit version), the Shell will swap itself out to one of
    the following:

    - Expanded memory
    - Extended memory
    - Disk (this is the slowest)

    The swapping is controlled by the shell internal command swap.  If
    swapping is enabled, the shell only uses 3K of memory whilst a child
    process is executing.

    Note: Swapping to Extended memory is probably the most dangerous unless
	  you have an XMS memory manager available.  The shell requires the
	  XMS manager to support the version 2 XMS specification.

2)  Command line editing has been added for non VI or EMACS mode.  A
    configuration file (sh.ini) allows the Command Line editing keys to be
    tailored to the users requirements.

3)  The command line prompt can be programmed to display 'useful' information.

4)  The shell uses Unix format file names (ie slashes and not backslashes) to
    delimit directories.  Some programs require certain environment variables
    to be in MS-DOS format (using backslashes).  The msdos or typeset -H
    commands allows these variables to be marked so that they are set correctly
    when the environment for a program is set up.

5)  The format of the command line which the shell passes to a program is
    configurable between normal, indirect files and environment variables
    (see SH.1).  A version of stdargv.c which supports the indirect file
    format (and wildcards from a normal command line) is included.

6)  Wild cards on drives (ie echo *:*.c will echo all the C files in the
    current directories of each drive) are supported.

7)  Full Interrupt 24 processing has been added.

8)  Alternate command interpreters are supported in shell scripts a la
    Unix V.4.

9)  A number of OS/2 internal commands (start, detach) have been addded.

10) The Notes file describes the fixes and changes between Release 2.1 and
    2.2

11) Significant parts of the shell have been re-written this release 2.1.


The following Bugs are known to exist in the Shell.

1)  See Notes file.

2)  EMACS line editing mode is not enabled or tested.

3)  I may have mis-interpreted the functionality of either the UNIX version
    of sh or ksh (tell me about it).

The shell was built using Microsoft C 5.1 and MASM v5.1 in large model mode
for MSDOS and Microsoft C6.0 for OS/2 (16 bit) and WATCOM 9 for both 32-bit
versions.

In addition, the version of open in your library must pass the O_NOINHERIT bit
on the MSDOS kernel.  The Microsoft C v5.1 library (and possibly the v6.0) does
not pass this bit on to the MSDOS open System call.  I fixed this using CodeView
to find where the library function masks off the bottom 2 bits.  Extracted the
object from the library and patched mask from 0x03 to 0x83 in the object and
reload into the library.  No Problem.  The Patch.Lib document describes the
process in more detail.

You can do want you like with this software as long as you don't sell it or
remove the Copyright notices in the sources or object.

If you have any problems or want to let me know about any enhancements you
have made (which could be included in a new general release), you can contact
me at

    Data Logic Limited
    Kings House
    Kymberley Way
    Harrow
    Middlesex, HA1 1YD
    UK.

    Phone : +44 1 863 0383
    E-Mail: istewart@datlog.co.uk

Note:
    Unix is a registered trademark of AT&T Bell Laboratories
    Microsoft, MSDOS, MASM and CodeView are registered trademarks of Microsoft
    Corporation
    OS/2 is a registered trademark of Internation Business Machines Corporation

Acknowledgements:

This program is based on ideas, code or parts of code developed by:

    David Korn and Steve Bourne (the original ideas)
    Charles Forsyth (original source for the MINIX Shell program)
    Erik Baalbergen (original source for the MINIX test program)
    Paul Falstad (original source for the maths functions from GNU zsh program)
    Simon J. Gerraty (the code for the new lexical analyser and the VI/EMACS
    edit functions).


CHANGES
Shell Version 2.2	CHANGES				December 1993

 MS-DOS SHELL - Copyright (c) 1990,3 Data Logic Limited and Charles Forsyth

 This code is based on (in part) the shell program written by Charles
 Forsyth and is subject to the following copyright restrictions:

 1.  Redistribution and use in source and binary forms are permitted
     provided that the above copyright notice is duplicated in the
     source form and the copyright notice in file sh6.c is displayed
     on entry to the program.

 2.  The sources (or parts thereof) or objects generated from the
     sources (or parts of sources) cannot be sold under any circumstances.

    $Header: /usr/users/istewart/src/shell/sh2.2/Release/RCS/Changes,v 2.2 1993/12/03 13:32:52 istewart Exp $

    $Log: Changes,v $
	Revision 2.2  1993/12/03  13:32:52  istewart
	Release 2.2

	Revision 2.1  1992/12/14  11:13:55  istewart
	BETA 215 Fixes and 2.1 Release

	Revision 2.0  1992/04/13  17:40:33  Ian_Stewartson
	MS-Shell 2.0 Baseline release

	
________________________________________________________________________________

We have standardised our MSDOS and Unix include files such that the same
include file exists our Unix system and MSDOS systems.  This makes porting
a lot easier.  If you don't want to edit the standard include files, you will
have to generate an new include file for the shell and include it as the first
include in all the C sources for the additional library functions and the
Shell itself.

These are the changes to the Microsoft C 5.1 and 6.0 include files.

The files changed are:

    limits.h - add missing definitions
    sys/stat.h - add missing definitions
    sys/types.h - add missing definitions

1.	limits.h

#ifdef OS2
#define NAME_MAX	63		/* Maximum file name length	*/
#define PATH_MAX	256		/* Maximum path name length	*/
#else
#define NAME_MAX	13		/* Maximum file name length	*/
#define PATH_MAX	128		/* Maximum path name length	*/
#endif

2.	sys/stat.h

#define S_ISDIR(m)	((((m) & S_IFMT) == S_IFDIR))
#define S_ISCHR(m)	((((m) & S_IFMT) == S_IFCHR))
#define S_ISREG(m)	((((m) & S_IFMT) == S_IFREG))
#define S_ISBLK(m)	((((m) & S_IFMT) == S_IFBLK))

extern mode_t _FAR_ _cdecl	umask (mode_t);

3.	sys/types.h

/*
 * Additional typedefs
 */

typedef unsigned char	bool;	/* Boolean: 0 = false, 1 = true		*/
typedef unsigned short	ushort;	/* 2-byte unsigned			*/
typedef ushort		u_short;
typedef ushort		mode_t;
typedef short		nlink_t;
typedef int		pid_t;
typedef ushort		uid_t;
typedef ushort		gid_t;
typedef short		nlink_t;

/*
 * System Constants
 */

#ifndef FALSE
#define FALSE	((bool)0)	/* Boolean 'false'			*/
#endif
#ifndef TRUE
#define TRUE	((bool)1)	/* Boolean 'true'			*/
#endif


SRC/SH6.C
/*
 * MS-DOS SHELL - Data Declarations
 *
 * MS-DOS SHELL - Copyright (c) 1990,3 Data Logic Limited and Charles Forsyth
 *
 * This code is based on (in part) the shell program written by Charles
 * Forsyth and is subject to the following copyright restrictions:
 *
 * 1.  Redistribution and use in source and binary forms are permitted
 *     provided that the above copyright notice is duplicated in the
 *     source form and the copyright notice in file sh6.c is displayed
 *     on entry to the program.
 *
 * 2.  The sources (or parts thereof) or objects generated from the sources
 *     (or parts of sources) cannot be sold under any circumstances.
 *
 *    $Header: /usr/users/istewart/src/shell/sh2.2/RCS/sh6.c,v 2.11 1993/11/09 10:39:49 istewart Exp $
 *
 *    $Log: sh6.c,v $
 *	Revision 2.11  1993/11/09  10:39:49  istewart
 *	Beta 226 checking
 *
 *	Revision 2.10  1993/08/25  16:03:57  istewart
 *	Beta 225 - see Notes file
 *
 *	Revision 2.9  1993/07/02  10:21:35  istewart
 *	224 Beta fixes
 *
 *	Revision 2.8  1993/06/14  11:00:12  istewart
 *	More changes for 223 beta
 *
 *	Revision 2.7  1993/06/02  09:52:35  istewart
 *	Beta 223 Updates - see Notes file
 *
 *	Revision 2.6  1993/02/16  16:03:15  istewart
 *	Beta 2.22 Release
 *
 *	Revision 2.5  1993/01/26  18:35:09  istewart
 *	Release 2.2 beta 0
 *
 *	Revision 2.4  1992/12/14  10:54:56  istewart
 *	BETA 215 Fixes and 2.1 Release
 *
 *	Revision 2.3  1992/11/06  10:03:44  istewart
 *	214 Beta test updates
 *
 *	Revision 2.2  1992/09/03  18:54:45  istewart
 *	Beta 213 Updates
 *
 *	Revision 2.1  1992/07/10  10:52:48  istewart
 *	211 Beta updates
 *
 *	Revision 2.0  1992/04/13  17:39:09  Ian_Stewartson
 *	MS-Shell 2.0 Baseline release
 *
 *
 */

#include <sys/types.h>
#include <sys/stat.h>
#include <stddef.h>
#include <stdio.h>
#include <setjmp.h>
#include <stdlib.h>
#include <limits.h>
#include <unistd.h>

#if defined (OS2) || defined (__OS2__)
#  define INCL_DOSSESMGR
#  include <os2.h>
#endif

#include "sh.h"

static char	*Copy_Right1 = "%s SH Version 2.2 (%s) - %s (OS %d.%d)\n";

#if (OS_VERSION == OS_OS2_32)
static char	*Copy_Right2 = "OS/2 (32-Bit)";
#elif (OS_VERSION == OS_OS2_16)
static char	*Copy_Right2 = "OS/2 (16-Bit)";
#elif (OS_VERSION == OS_DOS_32)
static char	*Copy_Right2 = "MS-DOS (32-Bit)";
#else
static char	*Copy_Right2 = "MS-DOS (16-Bit)";
#endif

static char	*Copy_Right3 = "Copyright (c) Data Logic Ltd and Charles Forsyth 1990, 93\n";
#if (OS_VERSION == OS_DOS_32)
bool		IgnoreHardErrors = FALSE;/* Hard Error Flag		*/
#endif
char		**ParameterArray = (char **)NULL; /* Parameter array	*/
int		ParameterCount = 0;	/* # entries in parameter array	*/
int		ExitStatus;		/* Exit status			*/
bool		ExpansionErrorDetected;
				/* interactive (talking-type wireless)	*/
bool		InteractiveFlag = FALSE;
bool		ProcessingEXECCommand;	/* Exec mode			*/
bool		UseConsoleBuffer = FALSE;/* Flag from dofc to		*/
					/* GetConsoleInput		*/
int		AllowMultipleLines;	/* Allow continuation onto	*/
					/* Next line			*/
int		Current_Event = 0;	/* Current history event	*/
bool		ChangeInitLoad = FALSE;	/* Change load .ini pt.		*/
unsigned int	GlobalFlags = 0;	/* Other global flags		*/
int		MaxNumberofFDs = 20;	/* Max no of file descriptors	*/

int		DisabledVariables = 0;	/* Disabled variables		*/
int		StartCursorPosition = -1;/* Start cursor position	*/

#if (OS_VERSION != OS_DOS)
unsigned int	SW_intr;		/* Interrupt flag		*/
bool		IgnoreInterrupts = FALSE;/* Ignore interrupts flag	*/
char		path_line[FFNAME_MAX];	/* Execution path		*/
#else
					/* Swap mode			*/
int		Swap_Mode = SWAP_EXPAND | SWAP_DISK;
#endif

Break_C		*Break_List;	/* Break list for FOR/WHILE		*/
Break_C		*Return_List;	/* Return list for RETURN		*/
Break_C		*SShell_List;	/* SubShell list for EXIT		*/
bool		RestrictedShellFlag = FALSE;	/* Restricted shell	*/
				/* History processing enabled flag	*/
bool		HistoryEnabled = FALSE;

void		*FunctionTree = (void *)NULL;	/* Function tree	*/
FunctionList	*CurrentFunction = (FunctionList *)NULL;
void		*AliasTree = (void *)NULL;	/* Alias tree		*/

#if (OS_VERSION != OS_DOS) && (OS_VERSION != OS_DOS_32)
void		*JobTree = (void *)NULL;	/* job tree		*/
bool		ExitWithJobsActive = FALSE;	/* Exit flag		*/
int		CurrentJob = 0;			/* No current		*/
int		PreviousJob = 0;		/* Previous Job		*/

/*
 * Session Info
 */

char		*SessionEndQName = (char *)NULL;	/* Queue	*/

/*
 * Special flag for EMX parameters
 */

bool		EMXStyleParameters = FALSE;
#endif

/*
 * redirection
 */

Save_IO		*SSave_IO;	/* Save IO array			*/
int		NSave_IO_E = 0;	/* Number of entries in Save IO array	*/
int		MSave_IO_E = 0;	/* Max Number of entries in SSave_IO	*/
S_SubShell	*SubShells;	/* Save Vars array			*/
int		NSubShells = 0;	/* Number of entries in SubShells	*/
int		MSubShells = 0;	/* Max Number of entries in SubShells	*/
int		LastNumberBase = -1;	/* Last base entered		*/

int		InterruptTrapPending;	/* Trap pending			*/
int		Execute_stack_depth;	/* execute function recursion	*/
					/* depth			*/
void		*VariableTree = (void *)NULL;	/* Variable dictionary	*/
VariableList	*CurrentDirectory;	/* Current directory		*/
char		*LastUserPrompt;	/* Last prompt output		*/
char		*LastUserPrompt1;	/* Alternate Last prompt output	*/
char		IFS[] = "IFS";		/* Inter-field separator	*/
char		PS1[] = "PS1";		/* Prompt 1			*/
char		PS2[] = "PS2";		/* Prompt 2			*/
char		PS3[] = "PS3";		/* Prompt 3			*/
char		PS4[] = "PS4";		/* Prompt 4			*/
char		PathLiteral[] = "PATH";
char		HomeVariableName[] = "HOME";
char		ShellVariableName[] = "SHELL";
char		HistoryFileVariable[] = "HISTFILE";
char		HistorySizeVariable[] = "HISTSIZE";
char		*ComspecVariable= "COMSPEC";
char		*ParameterCountVariable = "#";
char		*ShellOptionsVariable = "-";
char		*StatusVariable = "?";
char		SecondsVariable[] = "SECONDS";
char		RandomVariable[] = "RANDOM";
char		LineCountVariable[] = "LINENO";
char		*RootDirectory = "x:/";

#if (OS_VERSION != OS_DOS) && (OS_VERSION != OS_DOS_32)
char		WinTitleVariable[] = "WINTITLE";
#endif

char		*OldPWDVariable = "OLDPWD";
char		*PWDVariable = "PWD";
char		*ENVVariable = "ENV";

#if (OS_VERSION != OS_DOS) && (OS_VERSION != OS_DOS_32)
char		BATExtension[] = ".cmd";
#else
char		BATExtension[] = ".bat";
#endif

char		SHELLExtension[] = ".sh";
char		KSHELLExtension[] = ".ksh";
char		EXEExtension[] = ".exe";
char		COMExtension[] = ".com";
char		*NotFound = "not found";
char		*BasicErrorMessage = "%s: %s";
char		*DirectorySeparator = "/";
char		LastWordVariable[] = "_";
char		OptArgVariable[] = "OPTARG";
char		OptIndVariable[] = "OPTIND";
char		MailCheckVariable[] = "MAILCHECK";
char		FCEditVariable[] = "FCEDIT";
char		EditorVariable[] = "EDITOR";
char		VisualVariable[] = "VISUAL";
char		Trap_DEBUG[] = "~DEBUG";
char		Trap_ERR[] = "~ERR";
char		*LIT_NewLine = "\n";
char		*LIT_BadID = "bad identifier";
char		LIT_export[] = "export";
char		LIT_exit[] = "exit";
char		LIT_exec[] = "exec";
char		LIT_history[] = "history";
char		LIT_REPLY[] = "REPLY";
char		LIT_LINES[] = "LINES";
char		LIT_COLUMNS[] = "COLUMNS";
char		*ListVarFormat = "%s=%s\n";
char		*Outofmemory1 = "Out of memory";
char		*LIT_Emsg = "%s: %s (%s)";
char		*LIT_2Strings = "%s %s";
char		*LIT_3Strings = "%s %s%s";
char		*LIT_SyntaxError = "Syntax error";
char	 	*LIT_BadArray = "%s: bad array value";
char 		*LIT_ArrayRange = "%s: subscript out of range";
char		*LIT_BNumber = "[%d]";
char		*LIT_Invalidfunction = "Invalid function name";
char		LIT_Test[] = "[[";
int		MaximumColumns = 80;	/* Max columns			*/
int		MaximumLines = 25;	/* Max Lines			*/

/*
 * Fopen modes, different between IBM OS/2 2.x and the rest
 */

#ifdef __IBMC__
char		*OpenReadMode = "r";	/* Open file in read mode	*/
char		*OpenWriteMode = "w";	/* Open file in write mode	*/
char		*OpenAppendMode = "w+";	/* Open file in append mode	*/
char		*OpenWriteBinaryMode = "wb";/* Open file in append mode	*/
#else
char		*OpenReadMode = "rt";	/* Open file in read mode	*/
char		*OpenWriteMode = "wt";	/* Open file in write mode	*/
char		*OpenAppendMode = "wt+";/* Open file in append mode	*/
char		*OpenWriteBinaryMode = "wb";/* Open file in append mode	*/
#endif

#if (OS_VERSION != OS_DOS) && (OS_VERSION != OS_DOS_32)
STARTDATA	*SessionControlBlock;	/* Start a session info		*/
#endif

/*
 * Parser information
 */

char			CurrentLexIdentifier [IDENT+1];/* Identifier	*/
Source			*source;	/* yyparse/ScanNextToken source	*/
YYSTYPE			yylval;		/* result from ScanNextToken	*/
int			yynerrs;	/* Parse error flag		*/


/*
 * Global program mode information
 */

ExeMode 	ExecProcessingMode;	/* Global Program mode		*/

/*
 * Character Types array
 */

char		CharTypes [UCHAR_MAX + 1];

/*
 * Modified getopt values
 */

int		OptionIndex = 1;	/* optind			*/
int		OptionStart;		/* start character		*/
char		*OptionArgument;	/* optarg			*/

/*
 * Device directory.  The length of this string is defined by the variable
 * LEN_DEVICE_NAME_HEADER
 */

char		*DeviceNameHeader = "/dev/";

int		MemoryAreaLevel;/* Current allocation area		*/
long		flags = 0L;	/* Command line flags			*/
char		null[] = "";	/* Null value				*/
char		ConsoleLineBuffer[LINE_MAX + 1]; /* Console line buffer	*/

/*
 * Current environment
 */

ShellFileEnvironment	e = {
    (int *)NULL,		/* Error handler			*/
    0L,				/* IO Map for this level		*/
    (char *)NULL,		/* Current line buffer			*/
    (ShellFileEnvironment *)NULL,	/* Previous Env pointer		*/
};

/*
 * Some defines to print version and release info
 */

#define str(s)		# s
#define xstr(s)		str(s)

#if (OS_VERSION != OS_DOS) && (OS_VERSION != OS_DOS_32)
#  define OS_VERS_N	_osmajor / 10
#else
#  define OS_VERS_N	_osmajor
#endif

/*
 * The only bit of code in this module prints the version number
 */

void	PrintVersionNumber (FILE *fp)
{
    fprintf (fp, Copy_Right1, Copy_Right2, xstr (RELEASE), __DATE__,
	     OS_VERS_N, _osminor);
    fputs (Copy_Right3, fp);
}
